Improve OIDC multi-tenancy and providers docs (#34659)
quarkusio/quarkus: docs/src/main/asciidoc/security-openid-connect-providers.adoc

quarkus.oidc.credentials.secret=<Client Secret>
----

== Support for multiple providers

> If you would like to support authenticating users with more than one provider then a provider specific tenant configuration resolution has to be supported.

Reviewer: ```suggestion
If you would like to support authenticating users with more than one provider then a provider-specific tenant configuration resolution must be supported.
```

> Please read xref:security-openid-connect-multitenancy.adoc#static-tenant-resolution[Using OpenID Connect (OIDC) Multi-Tenancy] for more information.

Reviewer: ```suggestion
For more information, see the Quarkus xref:security-openid-connect-multitenancy.adoc#static-tenant-resolution[Using OpenID Connect (OIDC) Multi-Tenancy] guide.
```

== Access provider services with token propagation

> Sometimes, only authenticating users with a social provider is not enough, a provider specific service also needs to be accessed for the Quarkus OIDC `web-app` application to fetch some data from the provider service or even update some data in the provider service, on behalf of the currently authenticated user.

Reviewer: ```suggestion
Sometimes, only authenticating users with a social provider is not enough.
A provider-specific service also needs to be accessed for the Quarkus OIDC `web-app` application to fetch or update data from the provider service on behalf of the currently authenticated user.
```

> As mentioned in the xref:security-oidc-code-flow-authentication.adoc[OIDC code flow mechanism for protecting web applications] document, ID and access tokens are returned after the authorization code flow has been completed, with some providers like `GitHub` returning an access token only. It is this access token that has to be propagated to services such as `Google Calendar`, or `Spotify Playlists` for the currently authenticated user be able to use such services.

Reviewer: ```suggestion
As mentioned in the xref:security-oidc-code-flow-authentication.adoc[OIDC code flow mechanism for protecting web applications] guide, ID and access tokens are returned after the authorization code flow has been completed, with some providers like `GitHub` returning an access token only. 
It is this access token that has to be propagated to services such as `Google Calendar`, or `Spotify Playlists` for the currently authenticated user to be able to use such services.
```

> You do not have to bring provider specific libraries in order to achieve it, but use a reactive `Token Propagation` filter which can be bound to a given REST client with a simple annotation, please see xref:security-openid-connect-client-reference.adoc#token-propagation-reactive[Access token propagation] for more information.

Reviewer: ```suggestion
You do not have to bring provider-specific libraries in order to achieve this, but instead you can use a reactive `Token Propagation` filter, which can be bound to a given REST client with a simple annotation.
For more information, see the Quarkus xref:security-openid-connect-client-reference.adoc#token-propagation-reactive[Access token propagation] guide.
```

> For example, after you have configured <<google, Google provider>>, you can have events added to the user's Google Calendar with a REST client like this one:

Reviewer: ```suggestion
For example, after you have configured the <<google>> provider, you can have events added to the user's Google Calendar by using a REST client as shown in the following example:
```

[source,java]
----
package org.acme.calendar;

import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;

import io.quarkus.oidc.token.propagation.AccessToken;
import io.smallrye.mutiny.Uni;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;

@RegisterRestClient(configKey="google-calendar-api")

> @AccessToken

Reviewer: Would a call-out help here?

```suggestion
@AccessToken  <1>
```

@Path("/calendars/primary")
public interface GoogleCalendarClient {
    @POST
    @Path("events")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    Uni<String> addEvent(Event event);
    public static class Event {
        public String summary;
        public String kind = "calendar#event";
        public Time start;
        public Time end
    }

    public static class Time {
        public String dateTime;
        public String timeZone = "Europe/CET";
        public Time() {
        }
        public Time(String value) {
            dateTime = value;
	}
    }
}
----

> Note the `@AccessToken` annotation - it enables an access token propagation to the target service.

Reviewer: ```suggestion
<1> The `@AccessToken` annotation enables an access token propagation to the target service.
```

> For completeness, for this REST client to work, you need to configure the Google Calendar address and request that an access token is allowed to be used for accessing this service.

Reviewer: Remove fluff:
```suggestion
Finally,  you need to configure the Google Calendar address and request that an access token is always used for accessing this service, as outlined in the following example:
```

Author: @michelle-purcell Hi Michelle, `always` changes the meaning, specific scope add extra capabilities to issued tokens, which is why I used `allowed to be used for accessing`, let me try to make a more specific suggestion

Author: @michelle-purcell Does my suggestion look OK ?

Author: Let me apply it, and if you'd like to tweak it further, let me know please

Reviewer: > @michelle-purcell Does my suggestion look OK ?

Looks great :+1:  Just remove the superfluous space after `request`:  `request  the Google`
